Prototope/Behaviors.swift

```diff
@@ -0,0 +1,194 @@
+//
+//  Collisions.swift
+//  Prototope
+//
+//  Created by Saniul Ahmed on 07/02/2015.
+//  Copyright (c) 2015 Khan Academy. All rights reserved.
+//
+
+import Foundation
+
+//MARK: Public Behavior Types
+
+// Protocol for types describing the "configuration" of a behavior.
+public protocol BehaviorType {}
+
+// Value type describing the configuration of a collision behavior, specifying
+// the kind of the interaction (the collision conditions which should trigger the handler)
+// the layer with which the host layer is supposed to be colliding and a handler function
+public struct CollisionBehavior: BehaviorType {
+    public enum Kind {
+        case Entering
+        case Leaving
+    }
+    
+    let kind: Kind
+    let otherLayer: Layer
+    let handler: () -> Void
+    
+    public init(on kind: Kind, _ otherLayer: Layer, handler: ()->Void) {
+        self.kind = kind
+        self.otherLayer = otherLayer
+        self.handler = handler
+    }
+}
+
+//MARK: Behavior Bindings
+
+// Abstract. Describes a relationship between a layer and a single instance of a behavior.
+// Subclasses should encapsulate the state necessary to handle the behavior correctly.
+class BehaviorBinding: Equatable, Hashable {
```

Comment:
Could you make this a protocol (conforming to `Hashable`) instead of an abstract class? Those always feel kinda hacky. :)

Then `CollisionBehaviorBinding` could implement `==` with `===` rather than using a counter.